services/core/shared/core/blocks.js

```diff
@@ -0,0 +1,207 @@
+/*
+ * LiskHQ/lisk-service
+ * Copyright © 2020 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+const { Logger } = require('lisk-service-framework');
+
+const logger = Logger();
+
+const pouchdb = require('../pouchdb');
+const coreApi = require('./compat');
+const config = require('../../config');
+
+const indexList = [
+	'id',
+	'generatorPublicKey',
+	'generatorAddress',
+	'generatorUsername',
+	'height',
+	'numberOfTransactions',
+	'previousBlockId',
+	'unixTimestamp',
+	'totalAmount',
+	'totalFee',
+	'isFinal',
+	['height', 'isFinal'],
+	['generatorPublicKey', 'numberOfTransactions'],
+	['generatorAddress', 'numberOfTransactions'],
+	['generatorUsername', 'numberOfTransactions'],
+	['generatorPublicKey', 'totalAmount'],
+	['generatorAddress', 'totalAmount'],
+	['generatorUsername', 'totalAmount'],
+	['generatorPublicKey', 'unixTimestamp'],
+	['generatorAddress', 'unixTimestamp'],
+	['generatorUsername', 'unixTimestamp'],
+];
+
+let lastBlock = {};
+
+const getSelector = (params) => {
+	const result = {};
+	const selector = {};
+	if (params.height) selector.height = Number(params.height);
+	if (params.blockId) selector.id = String(params.blockId);
+	if (params.fromTimestamp) selector.unixTimestamp = { $gte: Number(params.fromTimestamp) };
+	if (params.toTimestamp) selector.unixTimestamp = { $lte: Number(params.toTimestamp) };
+	if (params.generatorPublicKey) selector.generatorPublicKey = String(params.generatorPublicKey);
+
+	if (params.limit) result.limit = Number(params.limit);
+	if (Number(params.offset) >= 0) result.skip = Number(params.offset);
+
+	// Not supported by the API but useful for caching
+	if (params.numberOfTransactions) selector.numberOfTransactions = {
+		$gte: Number(params.numberOfTransactions),
+	};
+	if (params.isFinal) selector.isFinal = params.isFinal;
+	if (params.isImmutable) selector.isFinal = params.isImmutable;
```

Comment:
`selector.isFinal` property is overriden by `params.isImmutable`. Kindly reverify.

Reply:
Thanks for pointing it out.